core/tracker/tracker.go

```diff
@@ -0,0 +1,252 @@
+// Copyright © 2022 Obol Labs Inc.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tracker
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/obolnetwork/charon/core"
+)
+
+//go:generate stringer -type=component
+
+// component refers to a core workflow component.
+type component int
+
+// Core components arranged in the order data flows through them.
+const (
+	scheduler component = iota
+	fetcher
+	consensus
+	validatorAPI
+	parSigDBInternal
+	parSigEx
+	parSigDBThreshold
+	sigAgg
+)
+
+// event represents an event emitted by a core workflow component.
+type event struct {
+	duty      core.Duty
+	component component
+	pubkey    core.PubKey
+}
+
+// Tracker represents the component that listens to events from core workflow components.
+// It identifies where a duty gets stuck in the course of its execution.
+type Tracker struct {
+	input chan event
+
+	// events stores all the events in a particular duty.
+	events    map[core.Duty][]event
+	deadliner core.Deadliner
+	quit      chan struct{}
+}
+
+// NewTracker returns a new Tracker.
+func NewTracker(deadliner core.Deadliner) *Tracker {
+	t := &Tracker{
+		input:     make(chan event),
+		events:    make(map[core.Duty][]event),
+		quit:      make(chan struct{}),
+		deadliner: deadliner,
+	}
+
+	return t
+}
+
+// Run blocks and registers events from each component in tracker's input channel.
+func (t *Tracker) Run(ctx context.Context) error {
+	defer close(t.quit)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e := <-t.input:
+			// TODO(dhruv): do not store events for expired duties
+			t.events[e.duty] = append(t.events[e.duty], e)
+			t.deadliner.Add(e.duty)
+		case duty := <-t.deadliner.C():
+			failed, failedComponent, failedMsg := analyseFailedDuty(duty, t.events[duty])
+			if failed {
```

Comment:
you should probably call report in all cases, so you can test positive and negative cases. The report function can just ignore non-failed.